sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisReaderTest.java

```diff
@@ -131,42 +131,51 @@ public void readsThroughAllDataAvailable() throws IOException, TransientKinesisE
   @Test
   public void watermarkDoesNotChangeWhenToFewSampleRecords()
       throws IOException, TransientKinesisException {
-    Instant now = Instant.now();
-    Instant recordsStartTimestamp = now.minus(Duration.standardHours(1));
-    final long timestampMs = recordsStartTimestamp.getMillis();
-    Duration safetyPeriod = Duration.standardMinutes(1);
-    Instant kinesisWatermarkMinValue = now.minus(KinesisReader.MAX_KINESIS_STREAM_RETENTION_PERIOD);
+    try {
+      Instant now = Instant.now();
+      DateTimeUtils.setCurrentMillisFixed(now.getMillis());
+      Instant recordsStartTimestamp = now.minus(Duration.standardHours(1));
+      final long timestampMs = recordsStartTimestamp.getMillis();
+      Duration safetyPeriod = Duration.standardMinutes(1);
+      Instant minKinesisWatermark = now.minus(KinesisReader.MAX_KINESIS_STREAM_RETENTION_PERIOD);
 
-    prepareRecordsWithArrivalTimestamps(timestampMs, 1, KinesisReader.MIN_WATERMARK_MESSAGES / 2);
+      prepareRecordsWithArrivalTimestamps(timestampMs, 1, KinesisReader.MIN_WATERMARK_MESSAGES / 2);
 
-    for (boolean more = reader.start(); more; more = reader.advance()) {
-      assertThat(reader.getWatermark()).isBetween(
-          kinesisWatermarkMinValue.minus(safetyPeriod),
-          kinesisWatermarkMinValue.plus(safetyPeriod));
+      for (boolean more = reader.start(); more; more = reader.advance()) {
+        assertThat(reader.getWatermark()).isBetween(
+            minKinesisWatermark.minus(safetyPeriod),
+            minKinesisWatermark.plus(safetyPeriod));
+      }
+    } finally {
+      DateTimeUtils.setCurrentMillisSystem();
     }
   }
 
   @Test
-  @Ignore("https://issues.apache.org/jira/browse/BEAM-3317")
   public void watermarkAdvancesWhenEnoughRecordsReadRecently()
       throws IOException, TransientKinesisException {
-    Instant now = Instant.now();
-    Instant recordsStartTimestamp = now.minus(Duration.standardHours(1));
-    long timestampMs = recordsStartTimestamp.getMillis();
-    Duration safetyPeriod = Duration.standardMinutes(1);
-    Instant kinesisWatermarkMinValue = now.minus(KinesisReader.MAX_KINESIS_STREAM_RETENTION_PERIOD);
-
-    prepareRecordsWithArrivalTimestamps(timestampMs, 1, KinesisReader.MIN_WATERMARK_MESSAGES);
-
-    int recordsNeededForWatermarkAdvancing = KinesisReader.MIN_WATERMARK_MESSAGES;
-    for (boolean more = reader.start(); more; more = reader.advance()) {
-      if (--recordsNeededForWatermarkAdvancing > 0) {
-        assertThat(reader.getWatermark()).isBetween(
-            kinesisWatermarkMinValue.minus(safetyPeriod),
-            kinesisWatermarkMinValue.plus(safetyPeriod));
-      } else {
-        assertThat(reader.getWatermark()).isEqualTo(new Instant(timestampMs));
+    try {
+      Instant now = Instant.now();
+      DateTimeUtils.setCurrentMillisFixed(now.getMillis());
```

Comment:
This is OK with me. I think even better would be to inject a clock into the reader if it depends on such a service. I don't really know much about this connector, though, so it is just a vague suggestion.